contents/docs/self-host/configure/migrate-postgres-clickhouse.mdx

```diff
@@ -0,0 +1,102 @@
+---
+title: Migrate from Postgres to ClickHouse
+sidebar: Docs
+showTitle: true
+---
+
+> This documentation is in **Alpha**. Please provide feedback via the [PostHog Communty Slack workspace](/slack) or [via a GitHub issue](https://github.com/PostHog/posthog.com/issues).
+
+**PostHog backed by Postgres is now deprecated**. We will continue to provide support to Postgres backed installs but we strongly encourage you to migrate to PostHog backed by ClickHouse for improved performance, to receive new features, and continued support in the future.
+
+TODOs:
+- Confirm what will and will not be migrated
+
+## What will be migrated
+
+The following will be migrated:
+
+- Insights
+- Dashboards
+- Sessions
+- Annotations
+- Feature flags
+- Projects
+
+## What will not be migrated
+
+- Cohorts
+
+## Before you begin
+
+TODOs:
+- Is there a minimum version for PostHog + Postgres and a minimum version for PostHog + ClickHouse? For example, does the Replicator plugin have a minimum PostHog version to work within?
+
+In order to migrate you must have the following in place:
+
+1. The PostHog + Postgres installation
+2. A new PostHog + ClickHouse installation
+
+## Export
+
+TODOs:
+- Are these the correct tables to dump?
+- Should we be using `--data-only` to only dump the data and not the schema creation?
+
+The first step in the migration process is to create a Postgres database dump. With a PostHog + Postgres instance up and running, run the `pg_dump` command with the following format:
+
+```
+$ pg_dump -d [database] -h [host] -U [user] -T posthog_person -T posthog_persondistinctid -T posthog_event -T posthog_pluginlogentry -T posthog_sessionrecordingevent --data-only -f [output-file]
+```
+
+For example:
+
+```
+$ pg_dump -d posthog -h localhost -U posthog -T posthog_person -T posthog_persondistinctid -T posthog_event -T posthog_pluginlogentry -T posthog_sessionrecordingevent --data-only -f posthog-migration.sql
+```
+
+The example command above creates a `posthog-migration.sql` that contains SQL statements to insert the data only, excluding event data, from the existing Postgres database.
+
+## Import
+
+The next step is to get the data into the new PostHog + ClickHouse installation. This installation still utilizes Postgres, but only for non-event data. Event data is now stored in ClickHouse.
+
+The import process has two steps:
+
+1. Run the exported SQL against the Postgres database for the new installation
+2. Use the PostHog replicator plugin to migrate event data from the old PostHog instance to the new one
+
+### Postgres import of non-event data
+
+Import the old Postgres data into the new PostHog Postgres database. The command should take the following format:
+
+```
+psql --dbname="postgres://[user]:[password]@[host]:[port]/[database]" < [input-file]
+```
+
+For example:
+
+```
+psql --dbname="postgres://postgres:postgres@localhost:5432/posthog" < posthog-migration.sql
```

Comment:
This still spits out a lot of errors and I'm unsure how to validate that the migration was successful.

Should Users be migrated? Should I still be presented with the preflight (because this is an entirely fresh install of PostHog) page? 

<details>
<summary>Click to expand</summary>
<pre>
<code>
kubectl exec -it -n posthog posthog-posthog-postgresql-0 -- psql --dbname="postgres://postgres:postgres@localhost:5432/posthog" <posthog-migration-2021-09-28-1100.sql
Unable to use a TTY - input is not a terminal or the right kind of file
SET
SET
SET
SET
SET
 set_config
------------

(1 row)

SET
SET
SET
SET
COPY 0
COPY 0
COPY 0
ERROR:  duplicate key value violates unique constraint "migrations_pkey"
DETAIL:  Key (id)=(1) already exists.
CONTEXT:  COPY migrations, line 1
COPY 0
ERROR:  duplicate key value violates unique constraint "django_content_type_pkey"
DETAIL:  Key (id)=(1) already exists.
CONTEXT:  COPY django_content_type, line 1
COPY 0
ERROR:  duplicate key value violates unique constraint "auth_permission_pkey"
DETAIL:  Key (id)=(1) already exists.
CONTEXT:  COPY auth_permission, line 1
ERROR:  null value in column "is_member_join_email_enabled" violates not-null constraint
DETAIL:  Failing row contains (017af29d-e39f-0000-8383-7a731cadcc6c, PostHog, 2021-07-29 14:14:31.598977+00, 2021-09-19 18:30:00.06962+00, {}, t, 9, f, {}, {}, null).
CONTEXT:  COPY posthog_organization, line 1: "017af29d-e39f-0000-8383-7a731cadcc6c	PostHog	2021-07-29 14:14:31.598977+00	2021-09-19 18:30:00.06962..."
ERROR:  duplicate key value violates unique constraint "posthog_team_pkey"
DETAIL:  Key (id)=(1) already exists.
CONTEXT:  COPY posthog_team, line 1
ERROR:  insert or update on table "posthog_user" violates foreign key constraint "posthog_user_current_organization_e8527570_fk_posthog_o"
DETAIL:  Key (current_organization_id)=(017af29d-e39f-0000-8383-7a731cadcc6c) is not present in table "posthog_organization".
COPY 0
COPY 0
COPY 0
ERROR:  duplicate key value violates unique constraint "django_migrations_pkey"
DETAIL:  Key (id)=(1) already exists.
CONTEXT:  COPY django_migrations, line 1
COPY 2
ERROR:  insert or update on table "posthog_eventdefinition" violates foreign key constraint "posthog_eventdefinition_team_id_818ed0f2_fk_posthog_team_id"
DETAIL:  Key (team_id)=(2) is not present in table "posthog_team".
COPY 0
ERROR:  duplicate key value violates unique constraint "posthog_propertydefinition_team_id_name_e21599fc_uniq"
DETAIL:  Key (team_id, name)=(1, $plugins_deferred) already exists.
CONTEXT:  COPY posthog_propertydefinition, line 123
COPY 0
COPY 0
COPY 0
ERROR:  insert or update on table "posthog_action" violates foreign key constraint "posthog_action_team_id_3a21e3a6_fk_posthog_team_id"
DETAIL:  Key (team_id)=(2) is not present in table "posthog_team".
ERROR:  insert or update on table "posthog_action_events" violates foreign key constraint "posthog_action_events_action_id_f6f1f077_fk_posthog_action_id"
DETAIL:  Key (action_id)=(8) is not present in table "posthog_action".
ERROR:  insert or update on table "posthog_actionstep" violates foreign key constraint "posthog_actionstep_action_id_b50d75e6_fk_posthog_action_id"
DETAIL:  Key (action_id)=(1) is not present in table "posthog_action".
ERROR:  insert or update on table "posthog_dashboard" violates foreign key constraint "posthog_dashboard_team_id_5c28ee1a_fk_posthog_team_id"
DETAIL:  Key (team_id)=(2) is not present in table "posthog_team".
ERROR:  null value in column "favorited" violates not-null constraint
DETAIL:  Failing row contains (11, Entered Free Trial -> Purchase (Premium), {"actions": [{"id": 8, "math": null, "name": "Installed App", "t..., null, null, f, 2, 4, {}, null, null, f, null, 2021-07-29 14:14:59.32907+00, null, f, null, f, cache_b8fb379fa071003b90e50ea72774aca3, 0inQEB2P, null, null, null, null).
CONTEXT:  COPY posthog_dashboarditem, line 1: "11	Entered Free Trial -> Purchase (Premium)	{"actions": [{"id": 8, "math": null, "name": "Installed ..."
COPY 0
COPY 0
COPY 0
ERROR:  insert or update on table "posthog_elementgroup" violates foreign key constraint "posthog_elementgroup_team_id_3ced0286_fk_posthog_team_id"
DETAIL:  Key (team_id)=(2) is not present in table "posthog_team".
ERROR:  insert or update on table "posthog_element" violates foreign key constraint "posthog_element_group_id_09134876_fk_posthog_elementgroup_id"
DETAIL:  Key (group_id)=(1) is not present in table "posthog_elementgroup".
COPY 0
COPY 0
COPY 0
ERROR:  insert or update on table "posthog_organizationmembership" violates foreign key constraint "posthog_organization_organization_id_6f92360d_fk_posthog_o"
DETAIL:  Key (organization_id)=(017af29d-e39f-0000-8383-7a731cadcc6c) is not present in table "posthog_organization".
COPY 0
ERROR:  duplicate key value violates unique constraint "posthog_plugin_pkey"
DETAIL:  Key (id)=(1) already exists.
CONTEXT:  COPY posthog_plugin, line 1
ERROR:  duplicate key value violates unique constraint "posthog_pluginconfig_pkey"
DETAIL:  Key (id)=(1) already exists.
CONTEXT:  COPY posthog_pluginconfig, line 1
COPY 1
COPY 0
COPY 0
COPY 0
COPY 0
COPY 0
COPY 0
COPY 0
COPY 0
COPY 0
COPY 0
COPY 0
COPY 0
 setval
--------
      1
(1 row)

 setval
--------
      1
(1 row)

 setval
--------
      1
(1 row)

 setval
--------
    188
(1 row)

 setval
--------
      1
(1 row)

 setval
--------
      1
(1 row)

 setval
--------
      1
(1 row)

 setval
--------
     47
(1 row)

 setval
--------
    214
(1 row)

 setval
--------
      1
(1 row)

 setval
--------
    766
(1 row)

 setval
--------
      8
(1 row)

 setval
--------
      8
(1 row)

 setval
--------
      1
(1 row)

 setval
--------
      1
(1 row)

 setval
--------
      1
(1 row)

 setval
--------
      4
(1 row)

 setval
--------
     15
(1 row)

 setval
--------
   1022
(1 row)

 setval
--------
     89
(1 row)

 setval
--------
   1727
(1 row)

 setval
--------
      1
(1 row)

 setval
--------
    174
(1 row)

 setval
--------
    176
(1 row)

 setval
--------
      4
(1 row)

 setval
--------
      2
(1 row)

 setval
--------
      4
(1 row)

 setval
--------
      1
(1 row)

 setval
--------
     62
(1 row)

 setval
--------
      1
(1 row)

 setval
--------
      1
(1 row)

 setval
--------
      2
(1 row)

 setval
--------
      1
(1 row)

 setval
--------
      1
(1 row)

 setval
--------
      1
(1 row)

 setval
--------
      1
(1 row)

 setval
--------
      1
(1 row)

 setval
--------
      1
(1 row)

 setval
--------
      1
(1 row)

 setval
--------
      1
(1 row)

 setval
--------
      1
(1 row)

 setval
--------
      1
(1 row)
</code>
</pre>
</details>

Reply:
A community member is using the following:

## export

```bash
$ pg_dump -d [database] -h [host] -U [user] \
--no-owner \
--no-privileges \
--exclude-table-data=posthog_action_events \
--exclude-table-data=posthog_cohortpeople \
--exclude-table-data=posthog_event \
--exclude-table-data=posthog_element \
--exclude-table-data=posthog_elementgroup \
--exclude-table-data=posthog_person \
--exclude-table-data=posthog_persondistinctid \
--exclude-table-data=posthog_pluginlogentry \
--exclude-table-data=posthog_sessionrecordingevent \
> export.sql
```

## import

```sql
$ psql -U postgres (enter password)

SELECT pg_terminate_backend(pid) FROM pg_stat_activity WHERE datname = 'posthog'; drop database posthog; create database posthog;

$ psql --dbname="postgres://postgres:postgres@localhost:5432/posthog" < export.sql
```